feat(tests): Improve failure diagnostics

Cobalt0s opened the pull request:
# Description

Custom comparators now collect **all** failure details instead of stopping at the first error, making test failures instantly actionable without a debugger. Each failure message pinpoints exactly what mismatched (field name, row index, expected vs actual values).

**Key improvements:**
- **Pretty print** -- prints essential messages, tabulated, focusing on root causes.
- **Continues on failure** -- collects ALL differences.
- **Precise locations** -- "Data[0].Fields[stagename]" not buried in deep diff.

# Unit Tests before and after
Examples of deliberately failing unit tests showing the dramatic improvement (modified `main` branch VS modified `test-utils` branch).

## ListObjectMetadata

### Before
<img width="1202" height="592" alt="image" src="https://github.com/user-attachments/assets/d5d8339e-04e9-45bd-8ff1-7b38147a0dbb" />
there is even more.

### After
<img width="923" height="541" alt="image" src="https://github.com/user-attachments/assets/f8ac5984-7a9d-41b2-b2af-78501e3b1239" />

## Read with associations

### Before
<img width="1193" height="823" alt="image" src="https://github.com/user-attachments/assets/a85e7d84-421e-4b1e-b8c7-92b1d5c7de1d" />

### After
<img width="1195" height="688" alt="image" src="https://github.com/user-attachments/assets/aaad428f-1f33-4adf-89a0-fae8197607f7" />


## Batch Write
### Before
<img width="1207" height="767" alt="image" src="https://github.com/user-attachments/assets/3209e1c0-e7ad-45f9-8170-1980ca9bd309" />

### After
<img width="1199" height="730" alt="image" src="https://github.com/user-attachments/assets/9ab26a98-55bf-4ece-9e88-63dd6dd1412b" />

## Write

### Before
<img width="1178" height="563" alt="image" src="https://github.com/user-attachments/assets/f4e364cf-2e49-4062-a479-4347c90d969d" />

### After
<img width="814" height="485" alt="image" src="https://github.com/user-attachments/assets/28f5c9e5-4913-45cc-a0e5-dc138be9a1c7" />

# PS
Even ListObjectMetadata without any custom comparators which uses the exact match comparison is displayed better, where one can see the path clearly:
<img width="1157" height="309" alt="image" src="https://github.com/user-attachments/assets/8b4ce4b0-f80b-47f2-9d69-355b269b6b01" />

Cobalt0s commented:
* **#2901** <a href="https://app.graphite.com/github/pr/amp-labs/connectors/2901?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#1644** <a href="https://app.graphite.com/github/pr/amp-labs/connectors/1644?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/amp-labs/connectors/1644?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Cobalt0s commented on providers/salesforce/read_test.go:
```diff
@@ -309,95 +308,6 @@ func TestRead(t *testing.T) { //nolint:funlen,gocognit,cyclop,maintidx
 }
 
-// comparatorSubsetReadWithAssociations extends ComparatorSubsetRead to also validate associations.
-func comparatorSubsetReadWithAssociations(serverURL string, actual, expected *common.ReadResult) bool {
```
Noticed that some assocaitions messages were duplicated (ex: statement 3 and 4 are the same as statement 5):
<img width="566" height="251" alt="Image" src="https://github.com/user-attachments/assets/69a7a879-8ac4-4f5e-8c66-9f5adb3d75ab" />

`comparatorSubsetReadWithAssociations` is a wrapper of `testroutines.ComparatorSubsetRead` and associations. But `testroutines.ComparatorSubsetRead` is now capable of handling associations internally so this is abolete and `testroutines.ComparatorSubsetRead` can be used directly.

Result of using `testroutines.ComparatorSubsetRead` directly:
<img width="494" height="181" alt="Image" src="https://github.com/user-attachments/assets/5ad1d9fa-5221-46b8-9dde-34ad483e014c" />

Cobalt0s reviewed (commented):
@RajatPawar the PR changes the utilities of the unit tests. The printing of failed tests is improved by returning not just the bool but the struct holding the status and list of messages reporting on differences between actual and expected.

AmandaYao00 commented on test/utils/mockutils/checks.go:
```diff
@@ -3,9 +3,82 @@ package mockutils
+		list[index] = fmt.Sprintf("\t❌ %v", text)
+	}
+
+	message := fmt.Sprintf("%v mistmatch:\n%v", dataName, strings.Join(list, "\n"))
```
```suggestion
	message := fmt.Sprintf("%v mismatch:\n%v", dataName, strings.Join(list, "\n"))
```

AmandaYao00 commented on test/utils/mockutils/batchWriteResult.go:
```diff
@@ -12,32 +14,40 @@ var BatchWriteResultComparator = batchWriteResultComparator{}
+			result.AddDiff(fmt.Sprintf("Result[%d] %s", i, diff))
 		}
+
+		result.AddMismatch(fmt.Sprintf("Result[%d].Success", i), expectedResult.Success, actualResult.Success)
```
AddMismatch signature is (dataName, gotData, expectedData). In some call sites, we're inverting it and passing in (expected, acutal) - which can potentially cause confusion in test results

Cobalt0s replied:
Hmmm, I will inspect the callers. This is critical.

AmandaYao00 commented on test/utils/testroutines/post-auth-info.go:
```diff
@@ -21,36 +21,29 @@ func (r PostAuthInfo) Run(t *testing.T, builder ConnectorBuilder[connectors.Auth
-				t.Fatalf("RawResoponse.Headers is not supported by PostAuthInfo TestCase")
-
-				return false
+				result.AddDiff("RawResoponse.Headers is not supported by PostAuthInfo TestCase")
```
```suggestion
				result.AddDiff("RawResponse.Headers is not supported by PostAuthInfo TestCase")
```

AmandaYao00 reviewed (approved):
Very cool!

Cobalt0s committed:
feat(tests): Improve difference reporting

Cobalt0s commented:
> Very cool!

Thanks!

Update:
- Renamed the `AddMismatch` function to `Assert`; it now serves both as an assertion and for storing mismatch messages.
- Removed conditional `if` statements before calls to `Assert`, since it already handles those checks internally.
- Updated the argument order to match `github.com/stretchr/testify/assert`: expected comes first, followed by actual.
- Ensured all call sites now use a consistent argument order.
